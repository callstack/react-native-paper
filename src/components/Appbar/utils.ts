import React from 'react';
import type { ColorValue, StyleProp, ViewStyle } from 'react-native';
import { StyleSheet, Animated } from 'react-native';

import overlay from '../../styles/overlay';
import { black, white } from '../../styles/themes/v2/colors';
import type { InternalTheme, ThemeProp } from '../../types';

export type AppbarModes = 'small' | 'medium' | 'large' | 'center-aligned';

export type AppbarChildProps = {
  isLeading?: boolean;
  color: string;
  style?: StyleProp<ViewStyle>;
};

const borderStyleProperties = [
  'borderRadius',
  'borderTopLeftRadius',
  'borderTopRightRadius',
  'borderBottomRightRadius',
  'borderBottomLeftRadius',
];

export const getAppbarBackgroundColor = (
  theme: InternalTheme,
  elevation: number,
  customBackground?: ColorValue,
  elevated?: boolean
) => {
  const { isV3, dark: isDarkTheme, mode, colors } = theme;
  const isAdaptiveMode = mode === 'adaptive';
  if (customBackground) {
    return customBackground;
  }

  if (!isV3) {
    if (isDarkTheme && isAdaptiveMode) {
      return overlay(elevation, colors?.surface);
    }

    return colors.primary;
  }

  if (elevated) {
    return theme.colors.elevation.level2;
  }

  return colors.surface;
};

export const getAppbarColor = ({
  color,
  isDark,
  isV3,
}: BaseProps & { color: string }) => {
  if (typeof color !== 'undefined') {
    return color;
  }

  if (isDark) {
    return white;
  }

  if (isV3) {
    return undefined;
  }

  return black;
};

export const getAppbarBorders = (
  style:
    | Animated.Value
    | Animated.AnimatedInterpolation<string | number>
    | Animated.WithAnimatedObject<ViewStyle>
) => {
  const borders: Record<string, number> = {};

  for (const property of borderStyleProperties) {
    const value = style[property as keyof typeof style];
    if (value) {
      borders[property] = value;
    }
  }

  return borders;
};

type BaseProps = {
  isDark: boolean;
  isV3: boolean;
};

type RenderAppbarContentProps = BaseProps & {
  children: React.ReactNode;
  shouldCenterContent?: boolean;
  isV3: boolean;
  renderOnly?: (string | boolean)[];
  renderExcept?: string[];
  mode?: AppbarModes;
  theme?: ThemeProp;
};

export const DEFAULT_APPBAR_HEIGHT = 56;
const MD3_DEFAULT_APPBAR_HEIGHT = 64;

export const modeAppbarHeight = {
  small: MD3_DEFAULT_APPBAR_HEIGHT,
  medium: 112,
  large: 152,
  'center-aligned': MD3_DEFAULT_APPBAR_HEIGHT,
};

export const modeTextVariant = {
  small: 'titleLarge',
  medium: 'headlineSmall',
  large: 'headlineMedium',
  'center-aligned': 'titleLarge',
} as const;

export const filterAppbarActions = (
  children: React.ReactNode,
  isLeading = false
) => {
  return React.Children.toArray(children).filter((child) => {
    if (!React.isValidElement<AppbarChildProps>(child)) return false;
    return isLeading ? child.props.isLeading : !child.props.isLeading;
  });
};

export const renderAppbarContent = ({
  children,
  isDark,
  shouldCenterContent = false,
  isV3,
  renderOnly,
  renderExcept,
  mode = 'small',
  theme,
}: RenderAppbarContentProps) => {
  return React.Children.toArray(children as React.ReactNode | React.ReactNode[])
    .filter((child) => child != null && typeof child !== 'boolean')
    .filter((child) =>
      // @ts-expect-error: TypeScript complains about the type of type but it doesn't matter
      renderExcept ? !renderExcept.includes(child.type.displayName) : child
    )
    .filter((child) =>
      // @ts-expect-error: TypeScript complains about the type of type but it doesn't matter
      renderOnly ? renderOnly.includes(child.type.displayName) : child
    )
    .map((child, i) => {
      if (
        !React.isValidElement<AppbarChildProps>(child) ||
        ![
          'Appbar.Content',
          'Appbar.Action',
          'Appbar.BackAction',
          'Tooltip',
        ].includes(
          // @ts-expect-error: TypeScript complains about the type of type but it doesn't matter
          child.type.displayName
        )
      ) {
        return child;
      }

      const props: {
        color?: string;
        style?: StyleProp<ViewStyle>;
        mode?: AppbarModes;
        theme?: ThemeProp;
      } = {
        theme,
        color: getAppbarColor({ color: child.props.color, isDark, isV3 }),
      };

      // @ts-expect-error: TypeScript complains about the type of type but it doesn't matter
      if (child.type.displayName === 'Appbar.Content') {
        props.mode = mode;
        props.style = [
          isV3
            ? i === 0 && !shouldCenterContent && styles.v3Spacing
            : i !== 0 && styles.v2Spacing,
          shouldCenterContent && styles.centerAlignedContent,
          child.props.style,
        ];
        props.color;
      }
      return React.cloneElement(child, props);
    });
};

const styles = StyleSheet.create({
  centerAlignedContent: {
    alignItems: 'center',
  },
  v2Spacing: {
    marginLeft: 8,
  },
  v3Spacing: {
    marginLeft: 12,
  },
});
