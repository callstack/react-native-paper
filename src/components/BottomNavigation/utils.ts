import type { InternalTheme } from 'src/types';

type BaseProps = {
  theme: InternalTheme;
};

export const getActiveTintColor = ({
  activeColor,
  theme,
}: BaseProps & {
  activeColor: string | undefined;
}) => {
  const {
    colors: { onSecondaryContainer },
  } = theme;
  if (typeof activeColor === 'string') {
    return activeColor;
  }

  return onSecondaryContainer;
};

export const getInactiveTintColor = ({
  inactiveColor,
  theme,
}: BaseProps & {
  inactiveColor: string | undefined;
}) => {
  const {
    colors: { onSurfaceVariant },
  } = theme;

  if (typeof inactiveColor === 'string') {
    return inactiveColor;
  }

  return onSurfaceVariant;
};

export const getLabelColor = ({
  tintColor,
  hasColor,
  focused,
  theme,
}: BaseProps & {
  tintColor: string;
  hasColor: boolean;
  focused: boolean;
}) => {
  const {
    colors: { onSurface, onSurfaceVariant },
  } = theme;

  if (hasColor) {
    return tintColor;
  }

  if (focused) {
    return onSurface;
  }

  return onSurfaceVariant;
};
