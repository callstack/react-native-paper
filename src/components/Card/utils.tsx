import type { StyleProp, ViewStyle } from 'react-native';

import color from 'color';

import { black, white } from '../../styles/themes/v2/colors';
import type { InternalTheme } from '../../types';

type CardMode = 'elevated' | 'outlined' | 'contained';

type BorderRadiusStyles = Pick<
  ViewStyle,
  Extract<keyof ViewStyle, `border${string}Radius`>
>;

export type CardActionChildProps = {
  compact?: boolean;
  mode?: string;
  style?: StyleProp<ViewStyle>;
};

export const getCardCoverStyle = ({
  theme,
  index,
  total,
  borderRadiusStyles,
}: {
  theme: InternalTheme;
  borderRadiusStyles: BorderRadiusStyles;
  index?: number;
  total?: number;
}) => {
  const { isV3, roundness } = theme;

  if (Object.keys(borderRadiusStyles).length > 0) {
    return {
      borderRadius: 3 * roundness,
      ...borderRadiusStyles,
    };
  }

  if (isV3) {
    return {
      borderRadius: 3 * roundness,
    };
  }

  if (index === 0) {
    if (total === 1) {
      return {
        borderRadius: roundness,
      };
    }

    return {
      borderTopLeftRadius: roundness,
      borderTopRightRadius: roundness,
    };
  }

  if (typeof total === 'number' && index === total - 1) {
    return {
      borderBottomLeftRadius: roundness,
    };
  }

  return undefined;
};

const getBorderColor = ({ theme }: { theme: InternalTheme }) => {
  if (theme.isV3) {
    return theme.colors.outline;
  }

  if (theme.dark) {
    return color(white).alpha(0.12).rgb().string();
  }
  return color(black).alpha(0.12).rgb().string();
};

const getBackgroundColor = ({
  theme,
  isMode,
}: {
  theme: InternalTheme;
  isMode: (mode: CardMode) => boolean;
}) => {
  if (theme.isV3) {
    if (isMode('contained')) {
      return theme.colors.surfaceVariant;
    }
    if (isMode('outlined')) {
      return theme.colors.surface;
    }
  }
  return undefined;
};

export const getCardColors = ({
  theme,
  mode,
}: {
  theme: InternalTheme;
  mode: CardMode;
}) => {
  const isMode = (modeToCompare: CardMode) => {
    return mode === modeToCompare;
  };

  return {
    backgroundColor: getBackgroundColor({
      theme,
      isMode,
    }),
    borderColor: getBorderColor({ theme }),
  };
};
