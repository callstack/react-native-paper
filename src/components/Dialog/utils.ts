import { StyleProp, ViewStyle } from 'react-native';

export type DialogChildProps = {
  style?: StyleProp<ViewStyle>;
};

export type DialogActionChildProps = DialogChildProps & {
  compact?: boolean;
  uppercase?: boolean;
};
