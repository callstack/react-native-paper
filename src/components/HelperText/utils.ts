import type { InternalTheme } from '../../types';

type BaseProps = {
  theme: InternalTheme;
  disabled?: boolean;
  type?: 'error' | 'info';
};

export function getTextColor({ theme, disabled, type }: BaseProps) {
  const {
    colors: { error, onSurfaceDisabled, onSurfaceVariant },
  } = theme;

  if (type === 'error') {
    return error;
  }

  if (disabled) {
    return onSurfaceDisabled;
  }

  return onSurfaceVariant;
}
