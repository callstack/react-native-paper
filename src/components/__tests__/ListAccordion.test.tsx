import * as React from 'react';
import { StyleSheet, View } from 'react-native';

import { render } from '@testing-library/react-native';
import color from 'color';

import { getTheme } from '../../core/theming';
import { Colors } from '../../styles/themes/tokens';
import ListAccordion from '../List/ListAccordion';
import ListAccordionGroup from '../List/ListAccordionGroup';
import ListIcon from '../List/ListIcon';
import ListItem from '../List/ListItem';
import { getAccordionColors } from '../List/utils';

const styles = StyleSheet.create({
  coloring: {
    color: Colors.error50,
  },
});

it('renders list accordion with children', () => {
  const tree = render(
    <ListAccordion
      left={(props) => <ListIcon {...props} icon="folder" />}
      title="Expandable list item"
    >
      <ListItem title="First Item" />
    </ListAccordion>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders multiline list accordion', () => {
  const tree = render(
    <ListAccordion
      title="Expandable list item"
      description="Describes the expandable list item"
    >
      <ListItem title="List item 1" />
    </ListAccordion>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders list accordion with left items', () => {
  const tree = render(
    <ListAccordion
      left={(props) => <ListIcon {...props} icon="star" />}
      title="Accordion item 1"
    >
      <ListItem
        left={(props) => <ListIcon {...props} icon="thumb-up" />}
        title="List item 1"
      />
    </ListAccordion>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders expanded accordion', () => {
  const tree = render(
    <ListAccordion title="Accordion item 1" expanded>
      <ListItem title="List item 1" />
    </ListAccordion>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders list accordion with custom title and description styles', () => {
  const tree = render(
    <ListAccordion
      title="Accordion item 1"
      description="Describes the expandable list item"
      titleStyle={styles.coloring}
      descriptionStyle={styles.coloring}
    >
      <ListItem title="List item 1" />
    </ListAccordion>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

describe('ListAccordion', () => {
  it('should not throw an error when id={0}', () => {
    const ListAccordionTest = () => (
      <ListAccordionGroup>
        <ListAccordion title="Testing list" id={0}>
          <View></View>
        </ListAccordion>
      </ListAccordionGroup>
    );

    expect(() => render(<ListAccordionTest />)).not.toThrow(
      'List.Accordion is used inside a List.AccordionGroup without specifying an id prop.'
    );
  });

  it('should throw an error when id={""}', () => {
    const ListAccordionTest = () => (
      <ListAccordionGroup>
        <ListAccordion title="Testing list" id={''}>
          <View></View>
        </ListAccordion>
      </ListAccordionGroup>
    );

    expect(() => render(<ListAccordionTest />)).toThrow(
      'List.Accordion is used inside a List.AccordionGroup without specifying an id prop.'
    );
  });
});

describe('getAccordionColors - description color', () => {
  it('should return theme color, for theme version 3', () => {
    expect(
      getAccordionColors({
        theme: getTheme(),
      })
    ).toMatchObject({
      descriptionColor: getTheme().colors.onSurfaceVariant,
    });
  });
});

describe('getAccordionColors - title text color', () => {
  it('should return theme color, for theme version 3', () => {
    expect(
      getAccordionColors({
        theme: getTheme(),
      })
    ).toMatchObject({
      titleTextColor: getTheme().colors.onSurface,
    });
  });

  it('should return primary color if it is expanded', () => {
    expect(
      getAccordionColors({
        theme: getTheme(),
        isExpanded: true,
      })
    ).toMatchObject({
      titleTextColor: getTheme().colors?.primary,
    });
  });
});

describe('getAccordionColors - ripple color', () => {
  it('should return theme color, for theme version 3', () => {
    expect(
      getAccordionColors({
        theme: getTheme(),
      })
    ).toMatchObject({
      rippleColor: color(getTheme().colors.onSurface)
        .alpha(0.12)
        .rgb()
        .string(),
    });
  });

  it('should return primary color if it is expanded', () => {
    expect(
      getAccordionColors({
        theme: getTheme(),
        isExpanded: true,
      })
    ).toMatchObject({
      rippleColor: color(getTheme().colors.primary).alpha(0.12).rgb().string(),
    });
  });
});
