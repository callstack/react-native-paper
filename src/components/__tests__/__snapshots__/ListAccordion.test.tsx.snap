// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders expanded accordion 1`] = `
<View>
  <View
    style={
      {
        "backgroundColor": "rgba(255, 251, 254, 1)",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": true,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "overflow": "hidden",
          },
          [
            {
              "paddingRight": 24,
              "paddingVertical": 8,
            },
            undefined,
          ],
        ]
      }
    >
      <View
        pointerEvents="none"
        style={
          [
            {
              "flexDirection": "row",
              "marginVertical": 6,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            [
              {
                "paddingLeft": 16,
              },
              {
                "flex": 1,
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontSize": 16,
                  },
                  {
                    "color": "rgba(103, 80, 164, 1)",
                  },
                  undefined,
                ],
              ]
            }
          >
            Accordion item 1
          </Text>
        </View>
        <View
          style={
            [
              {
                "marginVertical": 6,
                "paddingLeft": 8,
              },
              undefined,
            ]
          }
        >
          <Text
            accessibilityElementsHidden={true}
            importantForAccessibility="no-hide-descendants"
            pointerEvents="none"
            selectable={false}
            style={
              [
                {
                  "color": "rgba(73, 69, 79, 1)",
                  "fontSize": 24,
                },
                [
                  {
                    "lineHeight": 24,
                    "transform": [
                      {
                        "scaleX": 1,
                      },
                    ],
                  },
                  {
                    "backgroundColor": "transparent",
                  },
                ],
              ]
            }
          >
            chevron-up
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        false,
        [
          {
            "paddingRight": 24,
            "paddingVertical": 8,
          },
          undefined,
        ],
      ]
    }
  >
    <View
      style={
        [
          {
            "flexDirection": "row",
            "marginVertical": 6,
            "width": "100%",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          [
            {
              "paddingLeft": 16,
            },
            {
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "center",
            },
            undefined,
          ]
        }
        testID="undefined-content"
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              [
                {
                  "fontSize": 16,
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                },
                undefined,
              ],
            ]
          }
        >
          List item 1
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders list accordion with children 1`] = `
<View>
  <View
    style={
      {
        "backgroundColor": "rgba(255, 251, 254, 1)",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": false,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "overflow": "hidden",
          },
          [
            {
              "paddingRight": 24,
              "paddingVertical": 8,
            },
            undefined,
          ],
        ]
      }
    >
      <View
        pointerEvents="none"
        style={
          [
            {
              "flexDirection": "row",
              "marginVertical": 6,
            },
            undefined,
          ]
        }
      >
        <View
          pointerEvents="box-none"
          style={
            [
              {
                "alignItems": "center",
                "justifyContent": "center",
              },
              {
                "alignSelf": "center",
                "marginLeft": 16,
                "marginRight": 0,
                "marginVertical": 0,
              },
            ]
          }
        >
          <Text
            accessibilityElementsHidden={true}
            importantForAccessibility="no-hide-descendants"
            pointerEvents="none"
            selectable={false}
            style={
              [
                {
                  "color": "rgba(73, 69, 79, 1)",
                  "fontSize": 24,
                },
                [
                  {
                    "lineHeight": 24,
                    "transform": [
                      {
                        "scaleX": 1,
                      },
                    ],
                  },
                  {
                    "backgroundColor": "transparent",
                  },
                ],
              ]
            }
          >
            folder
          </Text>
        </View>
        <View
          style={
            [
              {
                "paddingLeft": 16,
              },
              {
                "flex": 1,
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontSize": 16,
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                  },
                  undefined,
                ],
              ]
            }
          >
            Expandable list item
          </Text>
        </View>
        <View
          style={
            [
              {
                "marginVertical": 6,
                "paddingLeft": 8,
              },
              undefined,
            ]
          }
        >
          <Text
            accessibilityElementsHidden={true}
            importantForAccessibility="no-hide-descendants"
            pointerEvents="none"
            selectable={false}
            style={
              [
                {
                  "color": "rgba(73, 69, 79, 1)",
                  "fontSize": 24,
                },
                [
                  {
                    "lineHeight": 24,
                    "transform": [
                      {
                        "scaleX": 1,
                      },
                    ],
                  },
                  {
                    "backgroundColor": "transparent",
                  },
                ],
              ]
            }
          >
            chevron-down
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders list accordion with custom title and description styles 1`] = `
<View>
  <View
    style={
      {
        "backgroundColor": "rgba(255, 251, 254, 1)",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": false,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "overflow": "hidden",
          },
          [
            {
              "paddingRight": 24,
              "paddingVertical": 8,
            },
            undefined,
          ],
        ]
      }
    >
      <View
        pointerEvents="none"
        style={
          [
            {
              "flexDirection": "row",
              "marginVertical": 6,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            [
              {
                "paddingLeft": 16,
              },
              {
                "flex": 1,
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontSize": 16,
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                  },
                  {
                    "color": "#f44336",
                  },
                ],
              ]
            }
          >
            Accordion item 1
          </Text>
          <Text
            numberOfLines={2}
            onTextLayout={[Function]}
            selectable={false}
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontSize": 14,
                  },
                  {
                    "color": "rgba(73, 69, 79, 1)",
                  },
                  {
                    "color": "#f44336",
                  },
                ],
              ]
            }
          >
            Describes the expandable list item
          </Text>
        </View>
        <View
          style={
            [
              {
                "marginVertical": 6,
                "paddingLeft": 8,
              },
              {
                "alignItems": "center",
                "height": 40,
                "justifyContent": "center",
              },
            ]
          }
        >
          <Text
            accessibilityElementsHidden={true}
            importantForAccessibility="no-hide-descendants"
            pointerEvents="none"
            selectable={false}
            style={
              [
                {
                  "color": "rgba(73, 69, 79, 1)",
                  "fontSize": 24,
                },
                [
                  {
                    "lineHeight": 24,
                    "transform": [
                      {
                        "scaleX": 1,
                      },
                    ],
                  },
                  {
                    "backgroundColor": "transparent",
                  },
                ],
              ]
            }
          >
            chevron-down
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders list accordion with left items 1`] = `
<View>
  <View
    style={
      {
        "backgroundColor": "rgba(255, 251, 254, 1)",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": false,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "overflow": "hidden",
          },
          [
            {
              "paddingRight": 24,
              "paddingVertical": 8,
            },
            undefined,
          ],
        ]
      }
    >
      <View
        pointerEvents="none"
        style={
          [
            {
              "flexDirection": "row",
              "marginVertical": 6,
            },
            undefined,
          ]
        }
      >
        <View
          pointerEvents="box-none"
          style={
            [
              {
                "alignItems": "center",
                "justifyContent": "center",
              },
              {
                "alignSelf": "center",
                "marginLeft": 16,
                "marginRight": 0,
                "marginVertical": 0,
              },
            ]
          }
        >
          <Text
            accessibilityElementsHidden={true}
            importantForAccessibility="no-hide-descendants"
            pointerEvents="none"
            selectable={false}
            style={
              [
                {
                  "color": "rgba(73, 69, 79, 1)",
                  "fontSize": 24,
                },
                [
                  {
                    "lineHeight": 24,
                    "transform": [
                      {
                        "scaleX": 1,
                      },
                    ],
                  },
                  {
                    "backgroundColor": "transparent",
                  },
                ],
              ]
            }
          >
            star
          </Text>
        </View>
        <View
          style={
            [
              {
                "paddingLeft": 16,
              },
              {
                "flex": 1,
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontSize": 16,
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                  },
                  undefined,
                ],
              ]
            }
          >
            Accordion item 1
          </Text>
        </View>
        <View
          style={
            [
              {
                "marginVertical": 6,
                "paddingLeft": 8,
              },
              undefined,
            ]
          }
        >
          <Text
            accessibilityElementsHidden={true}
            importantForAccessibility="no-hide-descendants"
            pointerEvents="none"
            selectable={false}
            style={
              [
                {
                  "color": "rgba(73, 69, 79, 1)",
                  "fontSize": 24,
                },
                [
                  {
                    "lineHeight": 24,
                    "transform": [
                      {
                        "scaleX": 1,
                      },
                    ],
                  },
                  {
                    "backgroundColor": "transparent",
                  },
                ],
              ]
            }
          >
            chevron-down
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders multiline list accordion 1`] = `
<View>
  <View
    style={
      {
        "backgroundColor": "rgba(255, 251, 254, 1)",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": false,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "overflow": "hidden",
          },
          [
            {
              "paddingRight": 24,
              "paddingVertical": 8,
            },
            undefined,
          ],
        ]
      }
    >
      <View
        pointerEvents="none"
        style={
          [
            {
              "flexDirection": "row",
              "marginVertical": 6,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            [
              {
                "paddingLeft": 16,
              },
              {
                "flex": 1,
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontSize": 16,
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                  },
                  undefined,
                ],
              ]
            }
          >
            Expandable list item
          </Text>
          <Text
            numberOfLines={2}
            onTextLayout={[Function]}
            selectable={false}
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontSize": 14,
                  },
                  {
                    "color": "rgba(73, 69, 79, 1)",
                  },
                  undefined,
                ],
              ]
            }
          >
            Describes the expandable list item
          </Text>
        </View>
        <View
          style={
            [
              {
                "marginVertical": 6,
                "paddingLeft": 8,
              },
              {
                "alignItems": "center",
                "height": 40,
                "justifyContent": "center",
              },
            ]
          }
        >
          <Text
            accessibilityElementsHidden={true}
            importantForAccessibility="no-hide-descendants"
            pointerEvents="none"
            selectable={false}
            style={
              [
                {
                  "color": "rgba(73, 69, 79, 1)",
                  "fontSize": 24,
                },
                [
                  {
                    "lineHeight": 24,
                    "transform": [
                      {
                        "scaleX": 1,
                      },
                    ],
                  },
                  {
                    "backgroundColor": "transparent",
                  },
                ],
              ]
            }
          >
            chevron-down
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
